xsnap/documentation/StaticModuleRecord.md

```diff
@@ -0,0 +1,181 @@
+# Static Module Records
+
+Revised: January 6, 2022
+
+Warning: These notes are preliminary. Omissions and errors are likely. If you encounter problems, please ask for assistance.
+
+## About
+
+A static module record prepares a module descriptor to be loaded and initialized as a new module into a compartment.
+
+> Compartments can use module descriptors directly instead of indirectly thru static module records.
+
+A static module record also provides access to the module bindings.
+
+## The StaticModuleRecord Constructor
+
+### StaticModuleRecord(descriptor)
+
+Returns a static module record object. Currently XS supports two kinds of module descriptor.
+
+### descriptor.source
+
+The `source` module descriptor constructs a static module record from a string that is parsed and compiled as a module. 
+
+	const smr = new StaticModuleRecord({ source: `
+		import x from "mod";
+		export let y = x;
+	` });
+
+Obviously, such module descriptor requires a JavaScript parser at runtime.
+
+### descriptor.bindings & descriptor.initialize
+
+The `bindings` & `initialize` module descriptor constructs a static module record from module bindings and a function to initialize the module.
+
+	const smr = new StaticModuleRecord({ 
+		bindings:[
+			{ import: "x", from: "mod" },
+			{ export: "y" },
+		], 
+		initialize($, meta) {
+			$.y = $.x;
+		} 
+	});
+	
+The module bindings allow the same expressiveness as import and export declarations without requiring a JavaScript parser at runtime. The `StaticModuleRecord` constructor checks the declarations and can throw a `SyntaxError`.
+
+Once the module is loaded and linked, the compartment calls the `initialize` function with two arguments:
+
+- `$`: the [module environment record](https://tc39.es/ecma262/#sec-module-environment-records),
+- `meta`: the module `import.meta`.
```

Comment:
I’ll be happy to get up to the point of having a meta to thread in the shim. The shim also threads the resolutions for the bindings, so they don’t have to be recomputed.